jerry-main/main-unix-snapshot.c

```diff
@@ -389,36 +374,195 @@ process_generate (cli_state_t *cli_state_p, /**< cli state */
-                                                                      is_save_literals_mode_in_c_format);
-    if (literal_buffer_size == 0)
+    switch (id)
     {
-      jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: Literal saving failed!\n");
-      return JERRY_STANDALONE_EXIT_CODE_FAIL;
+      case OPT_LITERAL_DUMP_HELP:
+      {
+        cli_help (prog_name_p, "literal-dump", literal_dump_opts);
+        return JERRY_STANDALONE_EXIT_CODE_OK;
+      }
+      case OPT_LITERAL_DUMP_FORMAT:
+      {
+        const char *fromat_str_p = cli_consume_string (cli_state_p);
+        if (!strcmp ("c", fromat_str_p))
+        {
+          is_c_format = true;
+        }
+        else if (strcmp ("list", fromat_str_p))
+        {
+          jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: Unsupported literal dump format.");
+          return JERRY_STANDALONE_EXIT_CODE_FAIL;
+        }
+        break;
+      }
+      case OPT_LITERAL_DUMP_OUT:
+      {
+        literals_file_name_p = cli_consume_string (cli_state_p);
+        break;
+      }
+      case CLI_OPT_DEFAULT:
+      {
+        const char *file_name_p = cli_consume_string (cli_state_p);
+
+        if (cli_state_p->error == NULL)
+        {
+          size_t size = read_file (input_pos_p, file_name_p);
+
+          if (size == 0)
+          {
+            return JERRY_STANDALONE_EXIT_CODE_FAIL;
+          }
+
+          snapshot_buffers[number_of_files] = (const uint32_t *) input_pos_p;
+          snapshot_buffer_sizes[number_of_files] = size;
+
+          number_of_files++;
+          const uintptr_t mask = sizeof (uint32_t) - 1;
+          input_pos_p = (uint8_t *) ((((uintptr_t) input_pos_p) + size + mask) & ~mask);
+        }
+        break;
+      }
+      default:
+      {
+        cli_state_p->error = "Internal error";
+        break;
+      }
     }
+  }
 
-    FILE *literal_file_p = fopen (literals_file_name_p, "wb");
+  if (check_cli_error (cli_state_p))
+  {
+    return JERRY_STANDALONE_EXIT_CODE_FAIL;
+  }
 
-    if (literal_file_p == NULL)
+  if (number_of_files < 1)
+  {
+    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: at least one input files must be passed.\n");
+    return JERRY_STANDALONE_EXIT_CODE_FAIL;
+  }
+
+  jerry_init (JERRY_INIT_EMPTY);
+
+  size_t lit_buf_sz = 0;
+  if (number_of_files == 1)
+  {
+    lit_buf_sz = jerry_get_literals_from_snapshot (snapshot_buffers[0],
+                                                   snapshot_buffer_sizes[0],
+                                                   literal_buffer,
+                                                   JERRY_BUFFER_SIZE,
+                                                   is_c_format);
+  }
+  else
+  {
+    /* The input contains more than one input snapshot file, so we must merge them first. */
+    const char *error_p = NULL;
+    size_t merged_snapshot_size = jerry_merge_snapshots (snapshot_buffers,
```

Comment:
Is this necessary? Cannot it be specified that literal dump works on one snapshot only? Doing merge in this subcommand seems to go against the 'do one thing but do that right' idea.

Reply:
IMHO, yes it is necessary. This is how we usually going to use it. E.g.: in IoT.js. It doesn't save the merged snapshot and does not modify the input, so I think it does only one thing (gather the literals). I don't see any problem with this.

Reply:
+1 for the auto-merge. Snapshot tool is not intended to run in low-end, so adding convenience functions should not hurt.

Reply:
side note: this code could become simpler if `jerry_merge_snapshots` could deal with corner cases like `number_of_files=1`. I recall that there was some discussion about this back when the snapshot tool was first implemented. one more item on a todo list?